libjxl: init at 2021-06-22-409efe02 (jpeg-xl reference implementation)

nh2 opened the pull request:
###### Motivation for this change

Packages the [JPEG XL](https://jpeg.org/jpegxl/)  reference implementation: [`jpeg-xl`](https://jpeg.org/jpegxl/).

I start this as a **draft PR** because:

* [x] need to fix that their git submodule setup is currently broken (https://gitlab.com/wg1/jpeg-xl/-/issues/57), so `src` is from my fork.
* [ ] I'm not sure I want to be maintainer for it, would be nice if somebody else would step up.

Also I'm PRing it against 20.09 to make it easy to test, of course it'll have to be rebased against `master` for the final PR.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/packages-looking-for-a-maintainer/5442/5

SuperSandro2000 commented on pkgs/development/libraries/jpeg-xl/default.nix:
```diff
@@ -0,0 +1,69 @@
+    sha256 = "0i58vr1jmawma4jck3v2ym76adksknj8jasnxvfj96zs2q5y025i";
+    # There are various submodules in `third_party/`.
+    fetchSubmodules = true;
+    # The `CMakeLists.txt` errors on absence of `.git` directories.
```
Can't we patch the cmakelists?

nh2 replied:
Would that be better? There may be legitimate use cases for keeping `.git`, e.g. if the build system uses it to determine `--version` output.

SuperSandro2000 replied:
If I remember this correctly ``.git`` is most of the time stripped because it can be really large and is not  really deterministic and the version is patched or passed in via a flag.

nh2 replied:
> If I remember this correctly `.git` is most of the time stripped because it can be really large and is not really deterministic and the version is patched or passed in via a flag. I think patching the version is better than keeping .git.

OK, then we need to patch it. I currently don't have time to look into it though (probably OK given that it's a draft PR anyway).

nh2 replied:
This seems to be addressed in the current version of `jpeg-xl`, it no longer errors on absence of `.git` directories.

SuperSandro2000 commented on pkgs/development/libraries/jpeg-xl/default.nix:
```diff
@@ -0,0 +1,69 @@
+    zlib
+  ];
+
+  cmakeFlags = [
+  ];
```
Is this meant to be empty?

nh2 replied:
I kept it empty for now just to make clear to whoever would like to step up for maintenance that this is a cmake bulid and that this is where options would go.

We might for example be interested in these: https://gitlab.com/wg1/jpeg-xl/-/blob/4d70bd58fbcb758b61446aba447fedd9177a00c9/CMakeLists.txt#L98-104

```cmake
set(JPEGXL_FORCE_SYSTEM_GTEST false CACHE BOOL
    "Force using system installed googletest (gtest/gmock) instead of third_party/googletest source.")
set(JPEGXL_FORCE_SYSTEM_BROTLI false CACHE BOOL
    "Force using system installed brotli instead of third_party/brotli source.")
set(JPEGXL_FORCE_SYSTEM_HWY false CACHE BOOL
    "Force using system installed highway (libhwy-dev) instead of third_party/highway source.")
```

if we want it to use our system libs (but I'm also OK with it using its submodule libs, as I'm not sure if jpeg-xl often patches them).

SuperSandro2000 replied:
I think we should use system libs if they are available or add a comment to use them in the future except if they are heavily patched.

nh2 replied:
Fine by me, but then we first need to have a good look into whether they changed important stuff or not.

nh2 replied:
I've added a comment.

SuperSandro2000 commented on pkgs/development/libraries/jpeg-xl/default.nix:
```diff
@@ -0,0 +1,69 @@
+  installCheckTarget = "test";
+
+  meta = with lib; {
+    homepage = http://libjpeg-turbo.virtualgl.org/;
```
```suggestion
    homepage = "https://libjpeg-turbo.virtualgl.org/";
```

nh2 replied:
Oops, I forgot to put in jpeg-xl's URL overall.

nh2 replied:
Done.

SuperSandro2000 commented on pkgs/development/libraries/jpeg-xl/default.nix:
```diff
@@ -0,0 +1,69 @@
+    homepage = http://libjpeg-turbo.virtualgl.org/;
+    description = "JPEG XL image format reference implementation.";
+    license = licenses.asl20;
+    maintainers = [ maintainers.nh2 ];
```
```suggestion
    maintainers = with maintainers; [ nh2 ];
```

nh2 replied:
Done.

SuperSandro2000 commented:
Result of `nixpkgs-review pr 103160` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>jpeg-xl</li>
  </ul>
</details>

SuperSandro2000 commented:
Result of `nixpkgs-review pr 103160` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>jpeg-xl</li>
  </ul>
</details>

```
hash mismatch in fixed-output derivation '/nix/store/05w2idsr1j896cjaxnd2hwiwk1sfq3hb-jpeg-xl-131953a':
  wanted: sha256:09vp73wqrhflv26xbikcpiw2bb37ffgc3skggpwycdcswn919m85
  got:    sha256:10xaqnmw48sz274m1gf2762xpy0sd453mmh10j8j432wx14y2pss
cannot build derivation '/nix/store/mrwwq6vfk4kikw7z8yjqx7zmrjj49a5j-jpeg-xl-2020-11-10-131953af.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/ama94bv01n0bh6fdk1a4l7ls6y5j3phj-env.drv': 1 dependencies couldn't be built
```

nh2 committed:
libjxl: init at 2021-06-22-409efe02 (jpeg-xl reference implementation)

nh2 commented:
Hmm, I just updated this, but now noticed that they moved from Gitlab to Github and renamed the project from `jpeg-xl` to `libjxl`:

https://gitlab.com/wg1/jpeg-xl/-/issues/245

nh2 commented:
fborg says

> @ofborg ofborg-eval — Complete, with errors

But it doesn't seem to post where I can read those errors.

SuperSandro2000 commented on pkgs/development/libraries/libjxl/default.nix:
```diff
@@ -0,0 +1,106 @@
+, libpng
+, libwebp
+, openexr
+, pkgconfig
```
```suggestion
, pkg-config
```

nh2 replied:
Done.

SuperSandro2000 commented on pkgs/development/libraries/libjxl/default.nix:
```diff
@@ -0,0 +1,106 @@
+    graphviz # for docs via doxygen component `dot`
+    doxygen # for docs
+    gtest
+    pkgconfig
```
```suggestion
    pkg-config
```

nh2 replied:
Done.

SuperSandro2000 commented on pkgs/development/libraries/libjxl/default.nix:
```diff
@@ -0,0 +1,106 @@
+stdenv.mkDerivation rec {
+
+  pname = "libjxl";
+  version = "2021-06-22-409efe02";
```
```suggestion
  version = "unstable-2021-06-22";
```

nh2 replied:
Done.

SuperSandro2000 commented on pkgs/development/libraries/libjxl/default.nix:
```diff
@@ -0,0 +1,106 @@
+    rev = "409efe027d6a4a4446b84abe8d7b2fa40409257d";
+    sha256 = "1akb6yyp2h4h6mfcqd4bgr3ybcik5v5kdc1rxaqnyjs7fp2f6nvv";
+    # There are various submodules in `third_party/`.
+    fetchSubmodules = true;
```
Those should be probably devondered.

nh2 replied:
See previous comment from https://github.com/NixOS/nixpkgs/pull/103160#discussion_r519525156

I have de-vendored `brotli`, but `libjxl` has many other submodules (https://github.com/libjxl/libjxl/tree/409efe027d6a4a4446b84abe8d7b2fa40409257d/third_party) that are frequently updated to pre-release versions, so I would want to put effort into that not before `libjxl` has become more stable.

I've added a code comment about that.

nh2 commented:
> > @ofborg ofborg-eval — Complete, with errors
> 
> But it doesn't seem to post where I can read those errors.

Apparently I had to look at the sibling github check `ofborg-eval-package-list-no-aliases`, which was failed, and contained:

```
warning: ignoring the user-specified setting 'restrict-eval', because it is a restricted setting and you are not a trusted user
error: while evaluating 'callPackageWith' at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-2/lib/customisation.nix:117:35, called from /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-2/pkgs/top-level/all-packages.nix:16622:12:
while evaluating 'makeOverridable' at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-2/lib/customisation.nix:67:24, called from /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-2/lib/customisation.nix:121:8:
anonymous function at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-2/pkgs/development/libraries/libjxl/default.nix:1:1 called without required argument 'pkgconfig', at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-2/lib/customisation.nix:69:16
```

According to @sternenseemann this means:

> it doesn't evaluate with aliases turned off because `pkgconfig` is not the proper package name anymore, but an alias. you should use `pkg-config` with a dash instead

nh2 commented:
`aarch64-linux` fails with: https://logs.nix.ci/?key=nixos/nixpkgs.103160&attempt_id=ccd5471f-9667-4950-87de-a3223fb2a8d1

```
In file included from /build/source/third_party/skcms/skcms.cc:2071:
/build/source/third_party/skcms/src/Transform_inl.h:619:25: internal compiler error: in instantiate_decl, at cp/pt.c:24602
  619 |     return cast<F>(lo|hi) * (1/65535.0f);
      |                         ^
Please submit a full bug report,
```

nh2 commented:
Ready to merge from my side.

For the internal compiler error failure on `aarch64-linux` failure in https://github.com/NixOS/nixpkgs/pull/103160#issuecomment-866388610, should I disable that arch, or leave it on so we can see if a future compiler upgrade will auto-fix it?

SuperSandro2000 commented:
> For the internal compiler error failure on `aarch64-linux` failure in [#103160 (comment)](https://github.com/NixOS/nixpkgs/pull/103160#issuecomment-866388610), should I disable that arch, or leave it on so we can see if a future compiler upgrade will auto-fix it?

Marking it broken saves hydra resources.

nh2 commented:
> Marking it broken saves hydra resources.

@SuperSandro2000 Thanks! Done.

SuperSandro2000 commented on pkgs/development/libraries/libjxl/default.nix:
```diff
@@ -0,0 +1,107 @@
+, zlib
+}:
+
+stdenv.mkDerivation rec {
+
+  pname = "libjxl";
```
```suggestion
stdenv.mkDerivation rec {
  pname = "libjxl";
```

SuperSandro2000 committed:
Update pkgs/development/libraries/libjxl/default.nix
